### ISSUE TITLE
Statement attribute on a case label suppresses 'bypassing variable initialization' error

### ISSUE THREAD

**Issue by dwblaikie** [llvm/llvm-project#84072]
https://godbolt.org/z/j4b69q1dq
```
void Func(int x) {
    switch (x) {
        [[likely]] case 0:
        case 1:
            int i = 3;
        case 2:
            break;
    }
    switch (x) {
        case 0:
        case 1:
            int i = 3;
        case 2:
            break;
    }
}
```
```
<source>:14:9: error: cannot jump from switch statement to this case label
   14 |         case 2:
      |         ^
<source>:13:17: note: jump bypasses variable initialization
   13 |             int i = 3;
      |              
```

**Comment by shafik** [comment]
So this changes the AST: https://godbolt.org/z/nev6hxT19 adding:

```console
|-AttributedStmt <line:4:9, line:6:22>
    |   | |-LikelyAttr <line:4:11>
```

before the `CaseStmt` I wonder if we are just not accounting for that.

**Comment by dwblaikie** [comment]
> So this changes the AST: https://godbolt.org/z/nev6hxT19 adding:
> 
> ```
> |-AttributedStmt <line:4:9, line:6:22>
>     |   | |-LikelyAttr <line:4:11>
> ```
> 
> before the `CaseStmt` I wonder if we are just not accounting for that.

Yep. Seems likely - I'd expect it's not a terribly difficult fix.

**Comment by shashu8660** [comment]
hii My name is shashank s I'm this year outreachy applicant . As  im beginner at open source i want to make my first contribution at good first  issue so i would like @dwblaikie to assign this issue to me

**Comment by dwblaikie** [comment]
Sure thing - done

**Comment by shashu8660** [comment]
im facing little difficulty to take my first step in this issue  @dwblaikie i would like you to guide me take the first step to solve this issue . I took so much time to install and learn the llvm work flow. plz i would like some assist to take first step to contribution.

**Comment by Rajveer100** [comment]
@dwblaikie @shafik 
Due to the `[[likely]]` attribute a nested structure is being formed in the AST, so as per the jump scope checker, we are skipping it:

```C++
// It's okay to jump out from a nested scope.
  if (CommonScope == ToScope) return;
```

**Comment by dwblaikie** [comment]
> im facing little difficulty to take my first step in this issue @dwblaikie i would like you to guide me take the first step to solve this issue . I took so much time to install and learn the llvm work flow. plz i would like some assist to take first step to contribution.

My usual approach when trying to figure this sort of thing out is to break at the diagnostic that's causing the problem (in this case the positive case has the diagnostic) - and have a new debugger session debugging the other case (non-diagnostic) and try to figure out where those things diverge.

Looks like @Rajveer100 might've pointed to a good/more precise place to start, though.

**Comment by shashu8660** [comment]
thank you @Rajveer100 and @dwblaikie  But i'm having different issue i know to how to solve the issue .I'm facing difficulty to locate the issue in llvm repo .This is my first contribution so be patience with me . and really sorry for asking easy problems . Can i get the last help from @dwblaikie or @Rajveer100 to locate the issue or a different approach

**Comment by dwblaikie** [comment]
> I'm facing difficulty to locate the issue in llvm repo 

Could you be more specific? Locating the code @Rajveer100 referred to? That's in: https://github.com/llvm/llvm-project/blob/49c3e78961f841717c0aeec0c04b6e5401ddd6ce/clang/lib/Sema/JumpDiagnostics.cpp#L965

**Comment by erichkeane** [comment]
Unassigning as @shashu8660 hasn't done anything on this in quite a while?  Feel free to ask for it back if you are working on it, else leaving it open to give others a chance to work on this.

**Comment by YutongZhuu** [comment]
Hi, I am a beginner with LLVM, I would like to make this as my first contribution to llvm. Thanks!

**Comment by YutongZhuu** [comment]
```cpp
// BuildScopeInformation(){
// ...
// Cases, labels, and defaults aren't "scope parents".  It's also
    // important to handle these iteratively instead of recursively in
    // order to avoid blowing out the stack.
    while (true) {
      Stmt *Next;
      if (SwitchCase *SC = dyn_cast<SwitchCase>(SubStmt))
        Next = SC->getSubStmt();
      else if (LabelStmt *LS = dyn_cast<LabelStmt>(SubStmt))
        Next = LS->getSubStmt();
      else if (AttributedStmt *AS = dyn_cast<AttributedStmt>(SubStmt))
        Next = AS->getSubStmt();
      else
        break;

      LabelAndGotoScopes[SubStmt] = ParentScope;
      SubStmt = Next;
    }
```

I can fix the problem by doing this, but I see AttributedStmt is handled as a special case in the function body:

```cpp
  case Stmt::AttributedStmtClass: {
    AttributedStmt *AS = cast<AttributedStmt>(S);
    if (GetMustTailAttr(AS)) {
      LabelAndGotoScopes[AS] = ParentScope;
      MustTailStmts.push_back(AS);
    }
    break;
  }
```

Can someone explain why can attrbutedStmt be a scope parent, and what the spacial case is doing?

@dwblaikie

**Comment by YutongZhuu** [comment]
I am not sure if I am asking stupid questions, apologize in advance~

**Comment by dwblaikie** [comment]
> // BuildScopeInformation(){
> // ...
> // Cases, labels, and defaults aren't "scope parents".  It's also
>     // important to handle these iteratively instead of recursively in
>     // order to avoid blowing out the stack.
>     while (true) {
>       Stmt *Next;
>       if (SwitchCase *SC = dyn_cast<SwitchCase>(SubStmt))
>         Next = SC->getSubStmt();
>       else if (LabelStmt *LS = dyn_cast<LabelStmt>(SubStmt))
>         Next = LS->getSubStmt();
>       +else if (AttributedStmt *AS = dyn_cast<AttributedStmt>(SubStmt))
>       + Next = AS->getSubStmt();
>       else
>         break;
> 
>       LabelAndGotoScopes[SubStmt] = ParentScope;
>       SubStmt = Next;
>     }
> I can fix the problem by doing this, but I see AttributedStmt is handled as a special case in the function body:
> 
>   case Stmt::AttributedStmtClass: {
>     AttributedStmt *AS = cast<AttributedStmt>(S);
>     if (GetMustTailAttr(AS)) {
>       LabelAndGotoScopes[AS] = ParentScope;
>       MustTailStmts.push_back(AS);
>     }
>     break;
>   }
> Can someone explain why can attrbutedStmt be a scope parent, and what the spacial case is doing?
> 
> [@dwblaikie](https://github.com/dwblaikie)

Do any tests fail with your change? Do any fail if you remove the special case for attributed statements above the loop?

**Comment by YutongZhuu** [comment]
> > // BuildScopeInformation(){
> > // ...
> > // Cases, labels, and defaults aren't "scope parents".  It's also
> > // important to handle these iteratively instead of recursively in
> > // order to avoid blowing out the stack.
> > while (true) {
> > Stmt *Next;
> > if (SwitchCase *SC = dyn_cast(SubStmt))
> > Next = SC->getSubStmt();
> > else if (LabelStmt *LS = dyn_cast(SubStmt))
> > Next = LS->getSubStmt();
> > +else if (AttributedStmt *AS = dyn_cast(SubStmt))
> > + Next = AS->getSubStmt();
> > else
> > break;
> > ```
> >   LabelAndGotoScopes[SubStmt] = ParentScope;
> >   SubStmt = Next;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I can fix the problem by doing this, but I see AttributedStmt is handled as a special case in the function body:
> > case Stmt::AttributedStmtClass: {
> > AttributedStmt *AS = cast(S);
> > if (GetMustTailAttr(AS)) {
> > LabelAndGotoScopes[AS] = ParentScope;
> > MustTailStmts.push_back(AS);
> > }
> > break;
> > }
> > Can someone explain why can attrbutedStmt be a scope parent, and what the spacial case is doing?
> > [@dwblaikie](https://github.com/dwblaikie)
> 
> Do any tests fail with your change? Do any fail if you remove the special case for attributed statements above the loop?

No, only expected fails failed. What do you mean by above the loop? Maybe you can comment on my PR so things are clearer. Btw, can I get assigned to this issue?

**Comment by dwblaikie** [comment]
> > > // BuildScopeInformation(){
> > > // ...
> > > // Cases, labels, and defaults aren't "scope parents".  It's also
> > > // important to handle these iteratively instead of recursively in
> > > // order to avoid blowing out the stack.
> > > while (true) {
> > > Stmt *Next;
> > > if (SwitchCase *SC = dyn_cast(SubStmt))
> > > Next = SC->getSubStmt();
> > > else if (LabelStmt *LS = dyn_cast(SubStmt))
> > > Next = LS->getSubStmt();
> > > +else if (AttributedStmt *AS = dyn_cast(SubStmt))
> > > 
> > > * Next = AS->getSubStmt();
> > >   else
> > >   break;
> > > 
> > > ```
> > >   LabelAndGotoScopes[SubStmt] = ParentScope;
> > >   SubStmt = Next;
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I can fix the problem by doing this, but I see AttributedStmt is handled as a special case in the function body:
> > > case Stmt::AttributedStmtClass: {
> > > AttributedStmt *AS = cast(S);
> > > if (GetMustTailAttr(AS)) {
> > > LabelAndGotoScopes[AS] = ParentScope;
> > > MustTailStmts.push_back(AS);
> > > }
> > > break;
> > > }
> > > Can someone explain why can attrbutedStmt be a scope parent, and what the spacial case is doing?
> > > [@dwblaikie](https://github.com/dwblaikie)
> > 
> > 
> > Do any tests fail with your change? Do any fail if you remove the special case for attributed statements above the loop?
> 
> No, only expected fails failed. 

Might be worth sending the patch then - perhaps it's all that's needed, but perhaps reviewers will have some ideas if there are missing test cases that would disprove this path.

> What do you mean by above the loop? Maybe you can comment on my PR so things are clearer. Btw, can I get assigned to this issue?

Oh, I see, you've already sent the PR, good stuff. I'll follow-up there.
